### DISCUSSION TITLE
ページテーブルエントリを作成する際の挙動・メモリレイアウトについて

### DISCUSSION THREAD

**Discussion by keisuke713** [nuta/microkernel-book#19] [Help] [answered]
### 概要

自分の認識が合っているか確認させて頂きたいのですが、
[こちら](https://github.com/nuta/microkernel-book/blob/main/kernel/riscv32/vm.c#L51)の処理は本書p71図6.4の1段目のページテーブル内のページテーブルエントリを作成する箇所に該当していて、[こちら](https://github.com/nuta/microkernel-book/blob/main/kernel/riscv32/vm.c#L55)の処理では `PTE_PADDR` が1段目のページテーブルエントリから2段目のページテーブルのメモリを算出している。であっていますでしょうか？？
また各ページテーブルエントリがp.72の図6.5が示すように物理アドレスでなく物理ページ番号を保持している理由は各ページ(テーブル)内でのオフセットは仮想アドレスのそれぞれ先頭10ビット・真ん中10ビット・下位12ビットが持っているから。で解釈合っていますでしょうか？

**Comment by nuta** [comment, marked as answer]
ページテーブルを辿る処理についてはどちらも正しいです 💯 `PTE_PADDR`は物理ページ番号から物理メモリアドレスへの変換をするマクロです。

なぜエントリが物理ページの先頭アドレスではなく、物理ページ番号を持っているのかは（おそらく）エントリ中のビット数の節約です。先頭アドレスをページの固定長を割ったより小さな数値を持っておけば良いので、節約できたところに他の情報（有効ビット・書き込み可ビットなど）を埋め込めます。

「おそらく」とはぐらかしたのは、CPUの設計次第で「物理アドレスをもつページテーブルエントリ」というのも実現できるはずだからです。ただ、x86-64含め基本的には物理ページ番号を持っていることが一般的だと思います。

**Reply by keisuke713** [reply]
> なぜエントリが物理ページの先頭アドレスではなく、物理ページ番号を持っているのかは（おそらく）エントリ中のビット数の節約です。先頭アドレスをページの固定長を割ったより小さな数値を持っておけば良いので、節約できたところに他の情報（有効ビット・書き込み可ビットなど）を埋め込めます。

なるほど、確かに節約できるならして他の情報を埋め込む方が効率が良いですね。

**Reply by keisuke713** [reply]
もう一点だけ確認させてください。
[こちら](https://github.com/nuta/microkernel-book/blob/main/servers/vm/page_fault.c#L73-L85)の処理で`uaddr` でなく `tmp_page` の仮想アドレスを介して物理アドレス( `paddr` )にセグメントの内容を読み込んでいるのは `uaddr` の値はvmのページテーブルに登録されていない値でページフォルトが起きてしまうから。と解釈したのですが合っていますでしょうか？

**Reply by nuta** [reply]
なぜなら`uaddr`の仮想アドレスは、そのコードを実行しているVMサーバのアドレス空間ではなく、ページフォルトが発生したプロセスのアドレス空間内を指しているためです。

そのため、`uaddr`にマップされている同じ物理ページにVMサーバからアクセスできるように、一旦VMサーバのアドレス空間の一部 (`tmp_page`) にその物理ページをマップしています。

**Reply by keisuke713** [reply]
認識合っていそうでよかったです！
ありがとうございます。
